### PR TITLE
Sanketdg/docfuse

### PR DESCRIPTION
**Pull request by SanketDG**
CC @Makman2

### CODE REVIEW

**Comment by sils1297** [comment]
@SanketDG why is this wip? Because of travis? We're removing travis right now.

File: setup.cfg [hunk 1]
```diff
@@ -14,7 +14,6 @@ norecursedirs =
     .env
 testpaths =
     coalib
-    docs
```
Comment (sils1297) [inline setup.cfg]: I'm against this: https://github.com/coala-analyzer/coala/issues/1790#issuecomment-219034544

**Comment by SanketDG** [comment]
@Makman2 upon much thinking I realized that the documentation parsing functions used for `FunctionMetadata` are fairly simple. The way we parse them right now, (as discussed on Gitter) is too complex for that usecase. So we have to find a way to transform the parsed docstring into a simple namedtuple.

Right now I am just creating a new function that parses the doc first, finds the relevant parts using `isinstance()` and then returns a namedtuple based on that.

Another thing we could do is to add a `simple: bool=False`(could use a better name) param to the function. In this way, when I pass `simple=True`,I could get a simple version of the parsed documentation.

Thoughts?

**Comment by sils1297** [comment]
@SanketDG the way I see the doc class thing: they are different classes and one of them should be renamed.

**Comment by Makman2** [comment]
Haven't we said we wanted a sequence/list of namedtuple?
And what's a "simple" version? :O

But I also agree with @sils1297, we should not merge stuff from `FunctionMetadata` and so on, they both deal with documentations, but the one is for general purpose documentation inside source code, while the other is purely Python specific and used in our coala core.
I think you should concentrate on getting the parsing function into the `coalib.bearlib.languages.documentation.DocumentationComment` only, and renaming the other one to maybe `PythonDocComment`/`PythonDocumentationComment` first^^

File: setup.cfg (hunk 1 above)
Comment (SanketDG) [inline setup.cfg]: CC @Makman2 @AbdealiJK @tushar-rishav @abhsag24 your thoughts on this?

File: tests/bearlib/languages/documentation/DocumentationCommentTest.py [hunk 2]
```diff
@@ -24,3 +24,69 @@ def test_fields(self):
...
+    def check_docstring(self,
+                        docstring,
+                        expected=[]):
+        self.assertIsInstance(docstring,
+                              str,
+                              "docstring needs to be a string for this test.")
+
+        self.assertIsInstance(expected,
+                              list,
+                              "docstring needs to be a list for this test.")
```
Comment (SanketDG) [inline tests/bearlib/languages/documentation/DocumentationCommentTest.py]: TODO needs to be expected

File: coalib/settings/DocstringMetadata.py [hunk 3]
```diff
@@ -24,54 +23,21 @@ def __init__(self, desc, param_dict, retval_desc):
...
-            if parse_mode == cls._ParseMode.RETVAL:
-                retval_desc = concat_doc_parts(retval_desc, line)
-            elif parse_mode == cls._ParseMode.PARAM:
-                param_dict[cur_param] = concat_doc_parts(param_dict[cur_param],
-                                                         line)
-            else:
-                desc = concat_doc_parts(desc, line)
-
-        return (cls(desc=desc.strip(),
+        for part in parsed_doc:  # pragma: no cover
```
Comment (SanketDG) [inline coalib/settings/DocstringMetadata.py]: Reason why this is not being covered:

By the current documentation parsing, there can be only one `Desc` section. When the loop passes through, it only finds it once and fails to go back to the for loop again. This causes a missed partial branch.

Need solutions :stuck_out_tongue_winking_eye: 

CC @Makman2

**Comment by sils1297** [comment]
unack

- https://github.com/coala-analyzer/coala/pull/2135/commits/eb51294974055e1ec11fc11eb94a5a2a81a3b42f as per discussion
- https://github.com/coala-analyzer/coala/pull/2135/commits/b00cc768507243226b340571d1bfb3ba775387a2 since it's just test duplication, that code is already in the docs where it belongs.

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2135/commits/c3d034eb9bbffabc625de3533069c237884fd4b5

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2135/commits/c03248078768993f1ba82847c4f55b70a7a87876 and https://github.com/coala-analyzer/coala/pull/2135/commits/ec3549809cfd95a86c861244ee4a61d9f67e1826

I don't think merging those classes is meaningful, you're moving a language independent internal routine over to another class where it isn't needed and then you use that method in the original location.

I'd really just take the one commit I acked from this PR and be done with it.

**Comment by sils1297** [comment]
unack c032480 and ec35498

**Comment by SanketDG** [comment]
reack 01fdf48

@sils1297 I will get a new PR for the new functions and all after some more refactoring

**Comment by SanketDG** [comment]
reack 01fdf48

**Comment by sils1297** [comment]
@rultor merge then

**Pull request merged**